[RF-DOCS] Rewrite the Layouts and Rendering guide (#57152)
rails/rails: guides/source/action_controller_overview.md

Reviewer: I'm wondering if we should start a bit simpler here for new comers that have never seen a controller.
Just render "Hello, world" as the first example, before diving into all the details of rendering. This would also avoid the confusion with `new` method till further down.
```suggestion
  def new
     render_to_string "Hello World!"
```

Or maybe just start with a simpler (non-crud) controller that renders plain (this would also require a routing code example, but that applies to the ClientsController as well):
````
```ruby
class WelcomeController < ApplicationController
  def greet
    render plain: "Hello World!"
  end
end
```
````

It would be nice if we could just call `#greet`, but that sadly doesn't work.
`WelcomeController.new.greet`.

Apparently we can call `.action`, which returns a Rack handler (I'm, not recommending we should use this 😄 ):
```
WelcomeController.action(:greet).call({}).last.body
=> "Hello World!"
```

Author: I'm not keen on this approach, I think we'd end up confusing the reader more than simplifying the concept. Non-crud actions is not something Rails recommends, so I don't think we should start a tutorial by teaching stuff that isn't recommended.

A little bit of confusion when trying to grasp a new concept, I think, is fine. 

That said, the `new` method confusion isn't ideal so I've reworked the example to use `index` instead which should be easier to understand.

Reviewer: The reworked section looks great.

Reviewer: I'm wondering if we should add a NOTE here to explaining the concept of an action?

Author: I don't think a `NOTE` is the right solution, but I agree we should explain a bit more. I've reworded the paragraph to say the request is matched to a "controller and action", and also linked to the routing guide so the reader can click through for further context.

I've also added the route to the below example which should provide further context. The definition and explanation of action should live in the routing guide I believe, so I don't want to add too much detail here.

What do you think?

Reviewer: I was looking at the "Getting Started" guide and it mentions "actions" but doesn't explain them.
The routing guide mentions "controller actions" a bunch but only has a bare explanation: "... it asks the router to match it to a controller action (aka method)".
Since these are named "controller actions" I would expect the "ActionController Overview" guide to describe them. I'll add this to the basecamp discussion about this guide.

Reviewer: Nevermind, I really like your explanation in the "ActionController Overview" guide!

updates model data accordingly.

### Structure of a Controller

A controller is a Ruby class which inherits from `ApplicationController` and has
methods just like any other class. Public methods in a controller are also known as
_actions_, as they are responsible for rendering responses.

```ruby
# app/controllers/posts_controller.rb

class PostsController < ApplicationController
  def index
  end
end
```

As per Rails conventions, controllers should define up to 7 conventional CRUD actions as demonstrated below. This convention is used by the [Router DSL](routing.html#crud-verbs-and-actions) to configure resourceful routes to the controller. Other actions may be defined and routed to, but these are outside Rails conventions. Further details are available in the [Routing guide](routing.html).

```ruby
# config/routes.rb

Rails.application.routes.draw do
  # A resourceful route to the `PostsController`
  resources :posts
end
```

```ruby
# app/controllers/posts_controller.rb

class PostsController < ApplicationController
  # GET /posts
  def index
    # Display all posts
  end

  # GET /posts/new
  def new
    # Render a form to create a new post
  end

>   # POST /posts

Reviewer: Having post refer to the request method and resource here, might confuse some new comers.
I also find `posts` somewhat abstract, not everyone know this refers to BlogPosts or NewsPosts.
If we are going to rename ClientsController, maybe use the more relatable ProductsController or BooksController?

Author: I used `Posts` to be keep with the "building a blog" theme in Rails tutorials, but I take your point here. I think `ProductsController` is a better example, will update.

Reviewer: The new "Getting Started" guide now builds a store and creates a ProductsController as well.

Author: Ah cool

method with the same name as the action.

NOTE: The `new` method in the above controller is an instance method, called on an instance of `PostsController`. This should not be confused with the `new` class method used to instantiate objects (`PostsController.new`).

### Naming Conventions

Rails favors pluralizing the resource in the controller's name. For example,
`PostsController` is preferred over `PostController` and
`SiteAdminsController` over `SiteAdminController` or `SitesAdminsController`.
However, the plural names are not strictly required — for example, `ApplicationController`.

Following this naming convention will allow you to
use [resourceful routes](routing.html#crud-verbs-and-actions) without
needing to qualify each with [additional options](routing.html#specifying-a-controller-to-use).
The convention also makes named route helpers consistent throughout your application.

The controller naming convention is different from models. While plural names
are preferred for controller names, the singular form is preferred for [model
names](active_record_basics.html#naming-conventions) (e.g. `Account` vs.
`Accounts`).

> Controller actions should be _public_, as only _public_ methods are callable as

Reviewer: Unrelated to this PR, but I think it would be useful to define in the guidelines when to use italic or bold for words.

Author: Yep, I can write a proposal in Basecamp of you wish?

actions. Helper methods within controllers which are _not_ intended to be actions
should be declared `private` or `protected`.

WARNING: Ensure that you do not override methods defined by `ActionController::Base`
when creating actions. Accidentally redefining them could result in `SystemStackError`.
If you limit your controllers to the 7 CRUD actions, this won't be an issue.

NOTE: If you must use a reserved method as an action name, one workaround is to
use a custom route to map the reserved method name to your non-reserved action
method.

### Rendering Responses

> Consider the below controller and route.

Reviewer: Nitpick, we first show the route, then the controller:
```suggestion
Consider the below route and controller.
```

end
```

```ruby
class PostsController < ApplicationController
  def index
  end
end
```

When a user navigates to `/posts`, Rails will create an instance
of `PostsController` and call its `index` method.
If the `index` method is empty, Rails will automatically render
`app/views/posts/index.html.erb`.

You can explicity define the template to render using the `render` method:

```ruby
class PostsController < ApplicationController
  def index
    # Renders `app/views/posts/feed.html.erb`
    render "feed"
  end
end
```

WARNING: Calling `render` does not `return` from the current scope. Statements after the method call will still be executed. Ensure you do not call `render` more than once in any given action, as it will raise an `AbstractController::DoubleRenderError`.

To enable rendering of additional formats, not just HTML, use a `respond_to` block. This will choose the correct template based on the `Accept` header in the HTTP request:

```ruby
class PostsController < ApplicationController
  def index
    respond_to do |format|
      format.html   # renders `app/views/posts/index.html.erb`
      format.xml    # renders `app/views/posts/index.xml.erb`
    end
  end
end
```

> See the [Layouts and Rendering guide](layouts_and_rendering.html) for futher details.

Reviewer: I think we can be more specific here and link to the section about rendering additional formats:
```suggestion
See the [Layouts and Rendering guide](layouts_and_rendering.html#multi-format-responses) for further details on rendering additional formats.
```

Author: This is actually a callout to that guide for rendering in general, not just rendering multi format responses. I'll make it clearer to convey that.

In the `index` method, the controller would typically create an array of the
`Post` model instances, and make it available as an instance variable called `@posts`
in the view:

```ruby
def index
  @posts = Post.all
end
```

NOTE: All controllers inherit from `ApplicationController`, which in turn
inherits from
[`ActionController::Base`](https://api.rubyonrails.org/classes/ActionController/Base.html).
For [API only](https://guides.rubyonrails.org/api_app.html) applications `ApplicationController` inherits from
[`ActionController::API`](https://edgeapi.rubyonrails.org/classes/ActionController/API.html).

### Redirecting Requests

Instead of rendering a fully-formed response, such as an HTML document, you might want to redirect the user to a different path. An [HTTP redirection status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status#redirection_messages) can be used for this. Rails' [`redirect_to`][] method uses `302 Found` by default.

```ruby
redirect_to photos_url
```

> This will send a `302` HTTP response to the browser with `photos_url` in the `Location` header. The browser will then make a new `GET` request to the `photos_url`.

Reviewer: Similar to the `head` method, maybe we can show an example redirect response here?

Author: Good shout, done!

Use the [`status:`](layouts_and_rendering.html#status) option to use a different HTTP response code. Both numeric and symbolic values are valid.

```ruby
redirect_to photos_path, status: :see_other
```

NOTE: When redirecting using `302 Found`, the client may not always make a `GET` request to the `Location`. JavaScript's `fetch` method will make a `GET` request after redirection from a `GET` or `POST` request. But, if the initial request is another method, for example `PUT`, it will use the same method when requesting the redirected location. This is not a problem in Rails because all [form submissions use `POST`](https://guides.rubyonrails.org/form_helpers.html#forms-with-patch-put-or-delete-methods), even when using [Turbo](http://turbo.hotwired.dev). <br><br> When making requests in custom JavaScript, redirect using [`303 See Other`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/303) to ensure a `GET` request to the redirected location, or [`307 Temporary Redirect`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/307) to preserve the original HTTP verb.

> NOTE: It's worth being aware that `redirect_to` doesn't move execution to a different method within the same request. It sends an HTTP response and then the browser makes a new request to the redirected location which won't have any context from the previous request.

Reviewer: This NOTE is easily missed with the Javascript one above it, and I think a lot of new comers run into the described issue.  (I'm also not a fan of how sequential NOTEs look).

Maybe we can move this up a bit higher, just above the `redirect_back` explanation?

Author: Done!

### Header-Only Responses

The [`head`][] method can be used to send responses with only headers to the browser. This is usually in response to an [HTTP `HEAD`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods/HEAD) request.

The `head` method accepts a number or symbol representing an HTTP status code.

```ruby
head :bad_request
# or
head 400
```

This would produce the following HTTP response:

```http
HTTP/1.1 400 Bad Request
Connection: close

> Date: Sun, 24 Jan 2010 12:15:53 GMT

Reviewer: We can probably update this date to make Rails seem a bit younger ;):
```suggestion
Date: Fri, 10 Juli 2026 12:15:53 GMT
```

Reviewer: Ah you removed it. Nice!

Author: Yeah, it wasn't sent as a header on my test Rails 8 app, so figured there's no point having it there!

Transfer-Encoding: chunked
Content-Type: text/html; charset=utf-8
X-Runtime: 0.013483
Set-Cookie: _blog_session=...snip...; path=/; HttpOnly
Cache-Control: no-cache
```

You can add additional HTTP headers if you wish.

```ruby
head :created, location: photo_path(@photo)
```

Which would produce:

```http
HTTP/1.1 201 Created
Connection: close

> Date: Sun, 24 Jan 2010 12:16:44 GMT

Reviewer: ```suggestion
Date: Fri, 10 Juli 2026 12:16:44 GMT
```

the `around` method must `yield` to execute the action.

This can be useful in cases that are more complex. As an example, you could rewrite
the `around_action` callback to measure execution time with a class:

```ruby
class ApplicationController < ActionController::Base
  around_action ActionDurationCallback
end

class ActionDurationCallback
  def self.around(controller)
    start_time = Time.now
    yield # This executes the action
    end_time = Time.now

    duration = end_time - start_time
    Rails.logger.info "Action #{controller.action_name} from controller #{controller.controller_name} took #{duration.round(2)} seconds to execute."
  end
end
```

In the above example, the `ActionDurationCallback`'s method is not run in the scope
of the controller but gets `controller` as an argument.

Request Parameters
------------------

Data sent by the incoming request is available in your controller using the
[`params`][] method which returns an [`ActionController::Parameters`][] object.

There are two types of parameter data:

- Query string parameters which are sent as part of the URL (for example, after
  the `?` in `http://example.com/accounts?filter=free`).

> - Parameters defined in the request body in non-`GET` requests,

Reviewer: To me this list would be more consistent if we list the two types in a similar way.
This is also how these parameters are referred to later on:
```suggestion
- Request body parameters in non-`GET` requests,
```

  usually from an HTML form.

Rails does not make a distinction between query string parameters and request body
parameters — both are available in the `params` object in your controller. For
example:

```ruby
class PostsController < ApplicationController
  # This action receives query string parameters from an HTTP GET request
  # at the URL "/post?filter=published"
  def index
    if params[:filter] == "published"
      @posts = Post.published
    else
      @posts = Post.all
    end
  end

  # This action receives parameters from a POST request to "/posts" URL with
  # form data in the request body.
  def create
    @posts = Post.new(params[:post])
    if @post.save
      redirect_to @post
    else
      render "new", status: :unprocessable_content
    end
  end
end
```

> NOTE: [`ActionController::Parameters`][] does not inherit from Hash, but it is similar in behavior. Although, unlike a Hash, symbolic and string keys, such as `:foo` and `"foo"`, are considered to be the same.

Reviewer: I kind of liked the original sentence using "mostly", as it is more corrrect. It also allows simplifying the second sentence.
Maybe:
```suggestion
NOTE: [`ActionController::Parameters`][] does not inherit from Hash, but it is mostly similar in behavior. Unlike a Hash, symbolic and string keys, such as `:foo` and `"foo"`, are considered to be the same.
```

Author: Updated as suggested! I didn't like the phrasing of the original sentence, and usually don't like vague terms like "mostly". In this case we point out a key difference in the second sentence so I think it works.

[`params`]:
    https://api.rubyonrails.org/classes/ActionController/StrongParameters.html#method-i-params
[`ActionController::Parameters`]:
    https://api.rubyonrails.org/classes/ActionController/Parameters.html

### Parameter Value Types

Values in an [`ActionController::Parameters`][] object must be a permitted scalar value defined in `ActionController::Parameters::PERMITTED_SCALAR_TYPES`.

```ruby
ActionController::Parameters::PERMITTED_SCALAR_TYPES
# => [String, Symbol, NilClass, Numeric, TrueClass, FalseClass, Date, Time, StringIO, IO, ActionDispatch::Http::UploadedFile, Rack::Test::UploadedFile]
```

> The object can contain nested hashes and array, but values within those must also contain a permitted scalar.

Reviewer: ```suggestion
The object can contain nested hashes and arrays, but values within those must also contain a permitted scalar.
```

```ruby
# config/routes.rb

Rails.application.routes.draw do
  # A resourceful route to the `ProductsController`
  resources :products
end
```

```ruby
# app/controllers/products_controller.rb

class ProductsController < ApplicationController
  # GET /products
  def index
    # Display all products
  end

  # GET /products/new
  def new
    # Render a form to create a new product
  end

  # POST /products
  def create
    # Handle the submission form rendered in `new`
    # and create the product in the database
  end

  # GET /products/:id
  def show
    # Show the product
  end

  # GET /products/:id/edit
  def edit
    # Render a form to edit the product
  end

  # PATCH /products/:id
  def update
    # Handle the submission form rendered in `edit`
    # and update the product in the database
  end

  # DELETE /product/:id
  def destroy
    # Delete the product
  end
end
```

> Once the [router matches](routing.html) an incoming request a controller and

Reviewer: Small typo I noticed:
```suggestion
Once the [router matches](routing.html) an incoming request to a controller and
```

In the `index` method, the controller would typically create an array of the
`Product` model instances, and make it available as an instance variable called `@products`
in the view:

```ruby
def index
  @products = Product.all
end
```

NOTE: All controllers inherit from `ApplicationController`, which in turn
inherits from
[`ActionController::Base`](https://api.rubyonrails.org/classes/ActionController/Base.html).
For [API only](https://guides.rubyonrails.org/api_app.html) applications `ApplicationController` inherits from
[`ActionController::API`](https://edgeapi.rubyonrails.org/classes/ActionController/API.html).

### Redirecting Requests

Instead of rendering a fully-formed response, such as an HTML document, you might want to redirect the user to a different path. An [HTTP redirection status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status#redirection_messages) can be used for this. Rails' [`redirect_to`][] method uses `302 Found` by default.

```ruby
redirect_to photos_url
```

This will send a `302` HTTP response to the browser with `photos_url` in the `Location` header. The browser will then make a new `GET` request to the `photos_url`.

> ```

Reviewer: We can use a http code block here to format it:
```suggestion
```http
```

Author: Ahhh nice, somehow missed the fact that we specify HTTP as a format!

> Given the above `ClientsController`, if a user goes to `/clients/new` in your

Reviewer: Somewhat out of scope, but I think we should add a header here.
Rendering responses is probably the most important function of a controller, it should at least be visible in the sidebar.
(this header should probably be moved up a bit depending on changes above).
```suggestion
### Rendering responses

Given the above `ClientsController`, if a user goes to `/clients/new` in your
```

Author: I agree with your sentiment here but I can't find an appropriate break in the prose to put in this heading. I think it would only disrupt the flow rather than add to it.

I also don't think "Creating a controller" is a very good section heading. We're not teaching how to generate a controller, it focuses more on structuring controllers, so maybe we should change the section heading to "Structuring Controllers and Rendering Responses"? 

Also I should add, this PR was focused on the Layouts and Rendering guide ... I tried to keep the changes to any other guides as minimal as possible, especially since they have already been updated. 

If you'd like I can spend some more time to tweak the action controller guide as well and maybe move stuff around.

application to add a new client, Rails will create an instance of
`ClientsController` and call its `new` method. If the `new` method is empty, Rails

> will automatically render `app/views/clients/new.html.erb` by default.

Reviewer: Somewhat out of scope, but I think this paragraph would be clearer as a numbered list:
```suggestion
Given the above `ClientsController`, if a user goes to `/clients/new` in your
application to add a new client, Rails will:

* create an instance of `ClientsController`
* call its `new` method
* if the `new` method is empty, automatically render `app/views/clients/new.html.erb` by default
```

Author: I honestly think it reads better as prose. I changed it to a list and it feels clunky to me. I've reworded it slightly anyway so let me know what you think

Reviewer: I like the reworded version.

    if @client.save
      redirect_to @client
    else
      render "new", status: :unprocessable_content
    end
  end

> end

Reviewer: This describes specific interaction between the controller and the view.
Maybe move this to the "Layouts and Rendering" guide, which we just linked to?

Author: This section is about processing parameters which I think belongs in the controller guide. I only added the `status` to this line as it's best practice (and needed for Turbo), so we should consistently use it across the guides. It is not, strictly speaking, relevant to the other changes.

</form>
```

When this form is submitted, the value of `params[:user]` will be:

> ```

Reviewer: ```suggestion
```ruby
```

{
  "name" => "Acme",
  "phone" => "12345",
  "address" => {
    "postcode" => "12345",
    "city" => "Carrot City"
  }
}
```

Note the nested hash in `params[:user][:address]`.

Rails provides helpers to construct HTML forms that adhere to Rails conventions. Refer
to the [Form Helpers guide](form_helpers.html) for further information.

### Composite Key Parameters

[Composite key parameters](active_record_composite_primary_keys.html) contain
multiple values in one parameter separated by a delimiter (such as an underscore).
Therefore, you will need to extract each value so that you can pass them to
Active Record. You can use the [`extract_value`][] method to do that.

> Consider the below controller and route:

Reviewer: I think mentioning both reads a bit easier than splitting it up 👍

{ "user" => { "name" => "acme", "address" => "123 Carrot Street" } }
```

#### Parameter Wrapping

Rails will automatically wrap parameters within a key denoting the corresponding resource,
and also add the controller name to JSON parameters.

For example, consider the below JSON object:

```json
{ "name": "acme", "address": "123 Carrot Street" }
```

If we send the above data to `create` action of the `UsersController`,
the JSON data will be wrapped within the `:user` key as:

```ruby
{
  controller: "users",
  action: "create",
  name: "acme",
  address: "123 Carrot Street",
  user: {
    name: "acme", address: "123 Carrot Street"
  }
}
```

NOTE: Rails adds a clone of the parameters to the hash within the key corresponding
to the resource's name. As a result, both the original version
of the parameters and the "wrapped" version of the parameters will exist in the
params object.

NOTE: While the action and controller names are available in the params object,
it is recommended to use the methods [`controller_name`][] and [`action_name`][] instead to access these values.

> Parameter wrapping is enabled by default, can be disabled using a configuration option:

Reviewer: This sentence seems to miss some words.
Maybe:
```suggestion
Parameter wrapping is enabled by default, but can be disabled using a configuration option:
```

Author: Done ... dyslexia wreaking havoc here!

```ruby
# config/application.rb

config.action_controller.wrap_parameters_by_default = false

> ```

Reviewer: Maybe just link to the configuration guide only. That guide should document how configuration can be set.

Author: That guide is a beast though, I think linking to it is fine when a detailed explanation is required, as that's the correct home for that kind of detail. If it's just a setting, I think keeping it in this guide is more reader friendly as they don't need to go elsewhere just for the syntax of a single setting.

You can also customize the name of the key or specific parameters you want to
wrap, see the [API documentation](https://api.rubyonrails.org/classes/ActionController/ParamsWrapper.html)
for more.

[`controller_name`]:
    https://api.rubyonrails.org/classes/ActionController/Metal.html#method-i-controller_name
[`action_name`]:
    https://api.rubyonrails.org/classes/AbstractController/Base.html#method-i-action_name

### Routing Parameters

Parameters specified as part of a route declaration in the `routes.rb` file are
also made available in the `params` hash. For example, we can add a route that

> captures the `:status` parameter for a posts:

Reviewer: ```suggestion
captures the `:status` parameter for a post:
```

Reviewer: As a side note, in this context `post` could also be interpreted as a POST request.

Author: Yeah I'll update this example to `product` along with the rest of the guide.

[`reset_session`]:
    https://api.rubyonrails.org/classes/ActionController/Metal.html#method-i-reset_session

### Session Stores

The storage mechanism for session data can be configured. By default, data is stored in an encrypted cookie, but other stores are available.

> All sessions have a unique ID representing the session object. Regardless of the chosen store, this ID is always stored in a cookie. The data can be stored using one of the following

Reviewer: Introducing the terms "session ID" and "session data" here makes this paragraph a bit clearer to me. We also use these terms exclusively for the rest of this section.
Maybe:
```suggestion
All sessions have a unique ID representing the session object. Regardless of the chosen store, this session ID is always stored in a cookie. The session data can be stored using one of the following
```

storage mechanisms:

> * [`ActionDispatch::Session::CookieStore`][] - Stores data in an encrypted cookie.

Reviewer: The other items use "Stores the data":
```suggestion
* [`ActionDispatch::Session::CookieStore`][] - Stores the data in an encrypted cookie.
```

  end
end
```

You may use different keys to assign different message types:

```ruby
flash[:notice]  = "Your post was created!"
flash[:alert]   = "Sorry, something when wrong when creating your post."
flash[:warning] = "Your post was created, but there were some problems."
```

Set a flash message when calling `redirect_to` by including it as a parameter:

```ruby
# Equivalent to setting flash[:notice]
redirect_to root_url, notice: "Your post was created!"

# Equivalent to setting flash[:alert]
redirect_to root_url, alert: "Sorry, something when wrong when creating your post."
```

Only `notice:` and `alert:` options may be used at the top-level. Storing messages under other keys needs the `flash:` option:

```ruby
# Equivalent to setting flash[:warning]
redirect_to root_url, flash: { warning: "Your post was created, but there were some problems." }
```

The flash does not render anything in the UI — it is a short-term data storage mechanism. Reading flash data and rendering the appropriate HTML is left up the the developer.

[`ActionDispatch::Flash::FlashHash`]:
    https://api.rubyonrails.org/classes/ActionDispatch/Flash/FlashHash.html

> #### Displaying flash messages

Reviewer: ```suggestion
#### Displaying Flash Messages
```

It is recommend to add code to render flash messages in your application layout, so
messages are automatically rendered on every page without additional steps.

Iterate through all the keys and values to render all set messages, and then you can use CSS to style the different types of messages based on their type:

```erb
<%# app/views/layouts/application.html.erb %>

<html>
  <%# ... %>
  <body>
    <% flash.each do |type, message| -%>
      <%= tag.div class: class_names("flash", type) do %>
        <p><%= message %></p>
      <% end %>
    <% end -%>

    <%# ... %>
    <%= yield %>
  </body>
</html>
```

#### `flash.keep` and `flash.now`

[`flash.keep`][] is used to carry over the flash value through to an additional
request. This is useful when there are multiple redirects.

For example, assume that the `root_url` routes to the `index` action in the controller below, and all requests here are redirected to `UsersController#index`.

> If an action sets the flash and redirects to `MainController#index`, those flash values will be lost when during the next redirect.

Reviewer: ```suggestion
If an action sets the flash and redirects to `MainController#index`, those flash values will be lost during the next redirect.
```
